ion-schema/src/isl/ranges.rs

```diff
@@ -345,16 +192,208 @@ impl WriteToIsl for U64Range {
     }
 }
 
-impl<T: PartialEq + Display> Display for Range<T> {
-    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
-        match &self.lower {
-            Limit::Closed(value) if self.lower == self.upper => value.fmt(f),
-            _ => {
-                f.write_str("range::[")?;
-                self.lower.fmt_for_display(f, "min")?;
-                f.write_str(",")?;
-                self.upper.fmt_for_display(f, "max")?;
-                f.write_str("]")
+/// This module contains the generic base for all of the "real" range types that we expose.
+mod base {
+    use super::*;
+
+    /// Trait to allow type-specific implementations to inject type-specific logic into the
+    /// constructor for [`Range`].
+    pub trait RangeValidation<T: PartialOrd> {
+        /// Checks if two limits would result in an empty range.
+        fn is_empty(start: &Limit<T>, end: &Limit<T>) -> bool {
+            match (start, end) {
+                (Limit::Inclusive(lower), Limit::Inclusive(upper)) => lower > upper,
+                (Limit::Exclusive(lower), Limit::Exclusive(upper))
+                | (Limit::Exclusive(lower), Limit::Inclusive(upper))
+                | (Limit::Inclusive(lower), Limit::Exclusive(upper)) => lower >= upper,
+                _ => false,
+            }
+        }
+    }
+
+    /// Checks if two limits would result in an empty range of integers. Returns true if there are
+    /// no integer values between `start` and `end`, taking into consideration the exclusivity of
+    /// the limits.
+    pub fn _is_int_range_empty<T: CheckedAdd + One + PartialOrd>(
```

Comment:
In Rust, names with a leading `_` indicate that the item is never used but you wanted to give it a meaningful name anyway. Unless there's another convention of which I'm unaware, consider renaming this.

Reply:
Oops... too much python recently. I'll fix it.